### PR TITLE
Add CIP on SET and REMOVE

### PR DESCRIPTION
**Pull request by sherfert**
https://github.com/sherfert/openCypher/blob/8051aa6b23d8e4b3e24d190d7b08ef796c52f090/cip/1.accepted/CIP2019-02-06-SET.adoc

### CODE REVIEW

**Comment by Mats-SX** [comment]
Build is red because of outdated license files. A rebase will probably fix it.

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+===== Multiple Assignments
+
+The `SET` clause can be followed by a comma-separated list of assignments.
+The order of the assignments is not relevant to the semantics of the clause.
+All assignments are evaluated atomically after all right-hand side expressions have been evaluated.
+Consider for example:
+[source, cypher]
+----
+MATCH (p1:Product{name:"laptop"}), (p2:Product{name:"tablet"})
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]:
```suggestion
MATCH (p1:Product {name: 'laptop'}), (p2:Product {name: 'tablet'})
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+SET p1.id = p2.id, p2.id = p1.id
+----
+
+All expressions are evaluated first and then assigned, which in this case means that `p1.id` and `p2.id` will swap their values.
+This will be unaffected by typing the clause in the opposite order: `SET p2.id = p1.id, p1.id = p2.id`
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]: Very nice attribute!

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+===== Assigning Multiple Properties
+
+Assigning multiple properties to the same node or relationship can be achieved in two ways:
+
+* Multiple assignments: `SET n.a = 5, n.b = 'word', n.c = $param`
+* Map augmentation: `SET n += {'a': 5, 'b': 'word', 'c': $param}`
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]: Incorrect Cypher.
```suggestion
* Map augmentation: `SET n += {a: 5, b: 'word', c: $param}`
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+===== Removing Properties
+
+The property graph model interprets a null property value as semantically identical to a missing or removed property.
+For this reason, there are multiple semantically identical ways to remove a property from a node or relationship:
+
+* Assign the property to null
+** `SET n.prop = null`
+** `SET n.prop = expression`  where expression evaluates to null, for example a variable or a parameter
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]: Personally, I think the last part of the sentence does not make it clearer
```suggestion
** `SET n.prop = expression` where expression evaluates to null
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+* Remove the property: `REMOVE n.prop`
+* Assign the node or relationship to a Map value that does not contain the previously existing property: `SET n = {a: var, b:var}`
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]:
```suggestion
* Assign to the node or relationship a map value that does not contain the previously existing property: `SET n = {a: var, b: var}`
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+* Adding multiple labels to a node: `SET n:Foo:Bar`
+* Removing one label from a node: `REMOVE n:Foo`
+* Removing multiple labels from a node: `REMOVE n:Foo:Bar`
+
+== Interaction with existing features
+
+=== Deleted Elements
+
+The `PropertyExpression` or `Variable` on the left-hand-side of a `SetItem` or `RemoveItem` might refer to a node or relationship that has been deleted previously in the same query.
+In this case, the `SET` or `REMOVE`  is a no-operation.
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]:
```suggestion
In this case, the `SET` or `REMOVE` is a no-operation.
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+=== Order Dependencies
+
+There are cases where the order of the incoming records can influence the final results.
+This could lead to unexpected behaviour.
+Consider the following example:
+[source, cypher]
+----
+MATCH (p1:Product {pid:85}), (p2:Product {pid:125})
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]:
```suggestion
MATCH (p1:Product {pid: 85}), (p2:Product {pid: 125})
```

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+This means the final result is a function of the order in which the `p2` nodes are found.
+This order is likely not something the user is aware of or in control of, and therefore could lead to surprising behaviour.
+
+==== Planner Warnings
+
+It is recommended that query planners detect cases where the query could result in order dependencies and issue warnings that describe this case to allow developers and users to take this into account when writing queries.
+The conditions to look for would be:
+
+* We have a `SET` on an element property where it is not known if the same element will occur on multiple rows or not.
+** If the elements come from a unique index, and their cardinality is never increased (eg. expand in pattern expression or cartesian product with another match), then the element is not repeated.
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]: I feel slightly uncomfortable talking about `unique index` in this context as indexes are not part of openCypher, but I don't have an alternative suggestion.

File: cip/1.accepted/CIP2019-02-06-SET.adoc [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+** If the user uses a `WITH DISTINCT`, or an aggregation that groups on the element before the set to ensure uniqueness, the element is not repeated.
+* And the right-hand side expression is not guaranteed to generate the same value for each case of the same left-hand side.
+** Literal expressions are safe.
+** Expressions that are only functions of the left-hand side element are safe.
+
+==== Runtime Errors
```
Comment (Mats-SX) [inline cip/1.accepted/CIP2019-02-06-SET.adoc]: It's a bit weird that this CIP allows three different levels of enforcement: no warnings, warnings, or errors, and that these are all listed as compliant implementations.

**Comment by hvub** [comment]
"This specification does not guarantee deterministic semantics if there are order dependencies." What are the precise reason for that design decision? Why can just the Planner Warnings rules of Section 4.2.1 be the general rules for a valid query, so that violation of these rule end in an error. I would argue that beyond these rules people to not now what they are doing anyhow. Or conversely, they exploit an unspecified behavior of a system for which they do not have any guarantees. What are the use cases that queries forbidden by the rules of Section 4.2.1 would fulfill? Why is not possible to or what is the pain of fulfilling these use case with other rule compliant queries?

**Comment by sherfert** [comment]
@hvub Are you suggesting making "Planner Warnings" mandatory instead of recommended? Sorry, but I am having some difficulties making sense of your comment.

**Comment by hvub** [comment]
No. I am asking for the reason why the rules behind "planning warning" (or rules of that kind) are not used to draw the hard line between valid SET clauses and non-valid SET clauses? Why let this whole problem slip into query runtime in the first place? Why not simply forbid SET clauses, where it is not statically clear that the assignment is save?

**Comment by sherfert** [comment]
During plan time you can find situations where you know it is save, situations where you know it is non-deterministic, and situations where you simply don't know. What do you suggest doing for that last category?

**Comment by hvub** [comment]
I do not try to provided an answer, I have ask a question. My question is: What is the effected use cases, the loss, the pain of defining the non-deterministic cases and the do-not-know cases as invalid queries? The CIP does not provide any motivation or explanation why these cases shall be allowed.

**Comment by sherfert** [comment]
Closing for now. Will reopen if work recommenced for this.

**Pull request closed without merging**